management_interface/management_interface/models.py

```diff
@@ -0,0 +1,56 @@
+import hashlib
+
+from django.db import models
+import uuid
+from datetime import datetime
+
+
+class BaseModel(models.Model):
+    """
+    Base model holding shared data
+    """
+    id = models.CharField(
+        primary_key=True, max_length=64, default=uuid.uuid4, editable=False, unique=True
+    )
+    created_at = models.DateTimeField(default=datetime.utcnow, editable=False)
+    updated_at = models.DateTimeField(auto_now=True, editable=False)
+
+
+class CareProviderLocation(BaseModel):
+    """
+    Model that represents a registered care provider location (branch) who wants to receive emails.
+    """
+    name = models.CharField(max_length=256, help_text="Helping Hands Wantage")
+    email = models.EmailField(null=False, db_index=True, unique=True, help_text="example@nhs.net")
+    ods_code = models.CharField(null=False, max_length=16, unique=True, help_text="VNJNK")
+    cqc_location_id = models.CharField(null=False, max_length=128, unique=True, help_text="1-11086090064")
+    registered_manager_id = models.ForeignKey("RegisteredManager", on_delete=models.CASCADE)
+
+
+class RegisteredManager(BaseModel):
+    """
+    Model that represents somebody who is a Care Quality Commission (CQC) registered manager.
+    """
+    given_name = models.CharField(max_length=256, db_index=True, help_text="Aislinn")
+    family_name = models.CharField(max_length=256, db_index=True, help_text="Mullen")
+    email = models.EmailField(null=False, db_index=True, unique=True, help_text="example@nhs.net")
+    cqc_registered_manager_id = models.CharField(max_length=128, help_text="1-XXXXXXXX")
+
+
+class CareRecipient(BaseModel):
+    """
+    Model that represents somebody who is receiving care from a care provider location
+    and who has had a HANS subscription made for them.
+    """
+
+    care_provider_location_id = models.ForeignKey("CareProviderLocation", on_delete=models.CASCADE)
+    nhs_number_hash = models.CharField(null=False, max_length=128, db_index=True, editable=False)
+    nhs_number = models.CharField(null=True, blank=True, max_length=32)
+    subscription_id = models.CharField(null=False, max_length=32, db_index=True, unique=True, editable=False)
+    provider_reference_id = models.CharField(null=False, max_length=128, db_index=True, unique=True, help_text="WANT45320482")
+
+    def save(self, *args, **kwargs):
```

Comment:
We're going to replace this anyway, but I think we need the `nhs_number` to be a write-only `@property` rather than calculating this on save.  Something like this:

```py
    
    def hash_nhs_number(self, nhs_number):
        # Probably worth parameterising these settings, but this demonstrates the call
        self.nhs_number_hash = hashlib.scrypt(nhs_number.encode('us-ascii'),
                                              salt="NaCl".encode('us-ascii'),
                                              n=1024, r=8, p=16, dklen=64)

    nhs_number = property(None, hash_nhs_number)

```

That way it'll properly blow up if you try to get `self.nhs_number` back out, and the hash value is available on an unsaved object.

If we do that, we don't need to implement `save()`.

Reply:
I spent quite some time playing with properties and it seems that they are in fact read-only in Django admin. You can show them in admin only if they are in `readonly_fields`. That also means we cannot make them input fields, however we need to. I ended up with overriding save() method.